feat: Implemented BeefyState

hMitov opened the pull request:
# Description
Implemented BeefyState class responsible for managing BEEFY finality rounds.
**Added essential fields for tracking:**
- Validator sets per session.
- Mandatory and non-mandatory block numbers for efficient finality.
- BEEFY genesis, finalization state, and next digest tracking.

**Introduced BeefySession class to store:**
- Authority sets for each session.
- A synchronized set of non-mandatory blocks that can be finalized for lower latency.

**P.S.**
Sessions are created by a validator set change in a block - we create a key value from the blockNumber and a session.

Non-mandatory blocks are added when votes occur.
We find the first session whose mandatory block number is greater than voted block and store that block in the previous to that session as a non-mandatory block.

Fixes #777 >

Commit:
feat: Implemented BeefyState class responsible for managing BEEFY finality rounds.

Zurcusa commented on src/main/java/com/limechain/beefy/dto/ValidatorSet.java:
```diff
@@ -1,12 +1,16 @@
 
 @Data
+@AllArgsConstructor
+@NoArgsConstructor
```
```suggestion
@Value
```

Value enforces private final fields. As we receive the validator set from a digest and do not change any data afterwards it is a suitable candidate to enforce this annotation.

Zurcusa commented on src/main/java/com/limechain/beefy/dto/SignedCommitment.java:
```diff
@@ -9,5 +9,5 @@
 public class SignedCommitment {
     private Commitment commitment;
-    private List<Optional<Signature>> signature;
+    private List<Optional<PubKey>> signatures;
```
```suggestion
    private List<Optional<byte[]>> signatures;
```

Zurcusa replied:
The signature is an array of bytes generated when we sign some data using a private key. The output is usually an array of bytes.

Grigorov-Georgi committed:
769 implement beefy networking (#794)

# Description

- Add all components needed to receive and send messages on the libp2p
substream "/beefy/2"
- Implement abstract classes for the network layer in order to reduce
code duplication

Fixes https://github.com/LimeChain/Fruzhin/issues/769

Zurcusa commented on src/main/java/com/limechain/beefy/dto/VoteMessage.java:
```diff
@@ -1,13 +1,11 @@
-    private BigInteger authorityId;
-    private Signature signature;
+    private byte[] authorityId;
+    private PubKey signature;
```
```suggestion
    private byte[] signature;
```

Zurcusa commented on src/main/java/com/limechain/beefy/dto/ValidatorSet.java:
```diff
@@ -1,12 +1,16 @@
 public class ValidatorSet {
-    private List<PubKey> validators;
-    private long id;
+    private List<byte[]> validators;
+    private BigInteger setId;
```
```suggestion
    List<byte[]> validators;
    BigInteger setId;
```

Zurcusa commented on src/main/java/com/limechain/beefy/state/BeefySession.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Set;
+import java.util.TreeSet;
+
+@Data
+public class BeefySession {
+    private ValidatorSet validatorSet;
+    private Set<BigInteger> nonMandatoryBlockNumbers = Collections.synchronizedSortedSet(new TreeSet<>());
```
```suggestion
@Value
public class BeefySession {
    ValidatorSet validatorSet;
    Set<BigInteger> nonMandatoryBlockNumbers = Collections.synchronizedSortedSet(new TreeSet<>());
```

Zurcusa commented on src/main/java/com/limechain/beefy/state/BeefyState.java:
```diff
@@ -0,0 +1,199 @@
+        );
+    }
+
+    private void saveBeefyFinalized() {
```
```suggestion
    private void persistBeefyFinalized() {
```

Zurcusa replied:
And for all other `save{}` methods.

Zurcusa commented on src/main/java/com/limechain/storage/DBConstants.java:
```diff
@@ -40,4 +40,9 @@ public class DBConstants {
+    //BeefyState keys
+    public static final String BEEFY_FINALIZED = "bs:beefyFinalized";
+    public static final String BEEFY_ROUND = "bs:beefyRound";
+    public static final String BEEFY_JUSTIFICATION = "bs:justification";
```
We should add a separate key for grandpa and beefy authorities.

Commit:
Merge branch '777-beefy-state' into dev

Commit:
fixed pr comments.

Commit:
Revert "fixed pr comments."

This reverts commit 83be1d6cc58533398f7891f53c684a7208c34349.

Commit:
fixed pr comments.

Commit:
Revert "fixed pr comments."

This reverts commit 270852dbf5b07f7d32a83a2af0f799963e83552e.

Commit:
fixed pr comments.
